src/library/features/crates/cratefeaturehelper.cpp

```diff
@@ -50,92 +50,27 @@ CrateId CrateFeatureHelper::createEmptyCrate() {
-        newCrate.setId(newCrateId);
-        m_pCrates->hierarchy().initClosureForCrate(newCrateId);
-        m_pCrates->hierarchy().generateCratePaths(newCrate);
-        qDebug() << "Created new crate" << newCrate;
-    } else {
-        DEBUG_ASSERT(!newCrateId.isValid());
-        qWarning() << "Failed to create new crate"
-                << "->"  << newCrate.getName();
-        QMessageBox::warning(
-                nullptr,
-                tr("Creating Crate Failed"),
-                tr("An unknown error occurred while creating crate: ") + newCrate.getName());
-    }
-    return newCrateId;
-}
-
-CrateId CrateFeatureHelper::createEmptySubrate(const Crate& parent) {
-    const QString proposedCrateName =
-            proposeNameForNewCrate(tr("New Subcrate"));
-    Crate newCrate;
-    while (!newCrate.hasName()) {
-        bool ok = false;
-        newCrate.parseName(
-                QInputDialog::getText(
-                        nullptr,
-                        tr("Create New Crate"),
-                        tr("Enter name for new crate:"),
-                        QLineEdit::Normal,
-                        proposedCrateName,
-                        &ok));
-        if (!ok) {
-            return CrateId();
-        }
-        if (!newCrate.hasName()) {
-            QMessageBox::warning(
-                    nullptr,
-                    tr("Creating Crate Failed"),
-                    tr("A crate cannot have a blank name."));
+              nullptr,
+              tr("Creating Crate Failed"),
+              tr("A crate cannot have a blank name."));
             continue;
         }
 
-        // select name from crateClosure join crates on id = childId group by childId having count(*) = 1;
-        // check if it has parent, if not compare with names from above
-        // in closure.
-        // Else split path into tokens and compare to those
-        if (m_pCrates->hierarchy().tokenizeCratePath(parent.getId()).contains(newCrate.getName())) {
+        if (m_pCrates->hierarchy().nameIsValidForHierarchy(newCrate.getName(), parent.getId())) {
             QMessageBox::warning(
-                    nullptr,
-                    tr("Creating Crate Failed"),
-                    tr("A crate by that name already exists."));
+              nullptr,
+              tr("Creating Crate Failed"),
+              tr("A crate by that name already exists."));
             newCrate.resetName();
             continue;
-        }
+            }
     }
 
     CrateId newCrateId;
     if (m_pCrates->insertCrate(newCrate, &newCrateId)) {
         DEBUG_ASSERT(newCrateId.isValid());
         newCrate.setId(newCrateId);
-        m_pCrates->hierarchy().initClosureForCrate(newCrateId);
-        if (m_pCrates->hierarchy().insertIntoClosure(parent.getId(), newCrateId)) {
-            m_pCrates->hierarchy().generateCratePaths(newCrate);
-            }
+        m_pCrates->hierarchy().addCrateToHierarchy(newCrate, parent);
```

Comment:
I don't think CrateHierarchy::addCrateToHierarchy should be accessed outside of CrateManager. The management of the hierarchy should be transparent to the rest of the application as much as practical. Add a new parameter to CrateManager::insertCrate to pass the parent Crate and have CrateManager::insertCrate call CrateHierarchy::addCrateToHierarchy.

Maybe it would make sense to change CrateManager::insertCrate to not take a const Crate but modify the passed Crate object to set its ID to the new ID instead of taking a CrateId* argument. It seems odd to me that is done in the GUI code here and that code is repeated in a few locations.